[JavaScript/TypeScript] Improve shebang and comments highlighting

deathaxe opened the pull request:
Fixes #2880

This PR implements  ...

1. shebang highlighting using the scheme of PHP/Java/ShellScript to ensure it being highlighted at the very first line, only.
2. xml highlighting in triple-slash comments in TS/TSX syntax according to Typescript specs, but with a more lazy/forgiving xml highlighting.
3. more forgiving comment punctuation highlighting in order to help ligatures for `//` or `///` to be displayed and to improve UX in case color schemes apply special highlighting to comment punctuation.
4. scopes trailing `//` in line comments `punctuation` for consistent highlighting.
5. adds various comment related test cases to illustrate changes and differences between syntaxes.

It feels odd to only highlight

-  `/*` and `*/` as punctuation in delimiters like `/***************************/` 
-  `//` in `/// comment` or `/////////////////////////////////////`

Note:

If `double-slash` and `triple-slash` sub-scopes are not needed, all line comment related contexts could probably be simplified/merged in some way.

wbond commented on JavaScript/JavaScript.sublime-syntax:
```diff
@@ -162,11 +170,51 @@ contexts:
+    - match: ^\#!
+      scope: punctuation.definition.comment.js
+      set: shebang-body
+    - match: ^|(?=\S)  # Note: Ensure to highlight shebang if Erlang is embedded.
```
Not sure I understand this comment about Erlang

deathaxe replied:
I guess it's just a copy and paste issue as this construct is used in several syntaxes.

wbond reviewed (dismissed):
Looks good to me. Only thing I wasn't sure about was the Erlang comment, which I am guessing was a copy-paste error?

deathaxe commented:
Fixed the typo. While doing so I found Java (PR) to also highlight trailing slashes in line comments. Added it as it might look better to have all slashes scoped the same in `// title ///////`.

Thom1729 commented on JavaScript/JavaScript.sublime-syntax:
```diff
@@ -109,8 +109,11 @@ variables:
+    - match: ''
+      push: [javascript, shebang]
 
+  javascript:
```
Maybe just `script` for the name of this context? Might be less confusing in the inherited syntaxes.

Thom1729 commented on JavaScript/tests/syntax_test_jsx.jsx:
```diff
@@ -1,5 +1,86 @@
 // SYNTAX TEST "Packages/JavaScript/JSX.sublime-syntax"
 
+// comment
```
Why duplicate the tests in other files?

deathaxe replied:
I just copied the whole bunch of comments from typescript to the other ones and adjusted rerequired stuff. May probably not needed in the ...X files at the moment, but it may not hurt to make sure not to regress something in case special treatment in react comments is needed in future.

I even prevented myself by regressing something by doing so (see the most recent force-pushes) ;-)

keith-hall commented:
> Fixed the typo. While doing so I found Java (PR) to also highlight trailing slashes in line comments. Added it as it might look better to have all slashes scoped the same in `// title ///////`.

Just a quick check - I recall that `wrap_line` functionality relies on the comment punctuation characters being scoped. Have you tested to ensure that nothing breaks when more characters in the comment receive these scopes (both leading and trailing punctuation)?

deathaxe commented on JavaScript/JavaScript.sublime-syntax:
```diff
@@ -121,33 +124,38 @@ contexts:
+  block-comments:
+    # empty block comments
+    - match: (/\*\*+/)
+      scope: comment.block.empty.java punctuation.definition.comment.java
```
Review test passed! 🤣

deathaxe commented on JavaScript/JavaScript.sublime-syntax:
```diff
@@ -121,33 +124,38 @@ contexts:
+
+  block-comments:
+    # empty block comments
+    - match: (/\*\*+/)
```
Yep. Resident of another quick test. Removed.

deathaxe commented:
Tried `wrap_line` in C and C# in SAFE MODE and it doesn't add another `//` when hitting enter at the end of a line comment. The only thing I've seen is `* ` being added  in block comments.

I have no strong opinion in applying such stuff to line comments. I just added it here, because there was another syntax (forgot which one), I had a discussion about different types of line comments, which ended up in exactly such rules.

Scoping trailing slashes or all is just to highlight "delimiting comments" which span the whole line, consistently. Maybe it's not worth it. Thoughts?

We should at least distinguish double-slash and triple-slash in a sane and consistent way as both have different meaning in TypeScript and most fonts provide ligatures for them.

keith-hall commented:
> Scoping trailing slashes or all is just to highlight "delimiting comments" which span the whole line, consistently. Maybe it's not worth it. Thoughts?

I personally like the idea, there may be color schemes which highlight a comment's punctuation differently from the comment text

> Which one looks better??

Consistency wins IMHO i.e. your current JavaScript :)

Thom1729 commented on JavaScript/JavaScript.sublime-syntax:
```diff
@@ -162,11 +170,54 @@ contexts:
+      push: line-comment-double-slash-body
+
+  line-comment-end:
+    - match: (//+)?\n
```
```suggestion
    - match: (?://+)?\n
```

deathaxe replied:
1. I am not sure about whether we should scope the newline character `punctuation`
2. This change would break syntax due to following `caputures 1`

Thom1729 replied:
Oops, my bad.

Thom1729 commented on JavaScript/JavaScript.sublime-syntax:
```diff
@@ -162,11 +170,54 @@ contexts:
+    - meta_include_prototype: false
+    - meta_scope: comment.line.shebang.js
+    # Note: Keep sync with first_line_match!
+    - match: \b(node|js)\b
```
```suggestion
    - match: \b(?:node|js)\b
```

deathaxe replied:
All shebang using syntaxes including `first_line_match` have this capture group. Changing it here would mean to also go though all the others as well. Not sure whether it's worth it as it is only the first line which is handled here. It doesn't have any effect on performance.

Thom1729 replied:
Makes sense.

deathaxe commented:
Thank's for having a look on it.
